astropy/uncertainty/builtin_distrs.py

```diff
@@ -0,0 +1,192 @@
+
+
+def normal(center, std=None, var=None, ivar=None, n_samples=1000,
+           dcls=Distribution, **kwargs):
+    """
+    Create a Gaussian/normal distribution.
+
+    Parameters
+    ----------
+    center : `~astropy.units.Quantity`
+        The center of this distribution
+    std : `~astropy.units.Quantity` or `None`
+        The standard deviation/σ of this distribution. Shape must match and unit
+        must be compatible with ``center``, or be `None` (if ``var`` or ``ivar``
+        are set).
+    var : `~astropy.units.Quantity` or `None`
+        The variance of this distribution. Shape must match and unit must be
+        compatible with ``center``, or be `None` (if ``std`` or ``ivar`` are set).
+    ivar : `~astropy.units.Quantity` or `None`
+        The inverse variance of this distribution. Shape must match and unit
+        must be compatible with ``center``, or be `None` (if ``std`` or ``var``
+        are set).
+    n_samples : int
+        The number of Monte Carlo samples to use with this distribution
+    dcls : class
+        The class to use to create this distribution.  Typically a
+        `Distribution` subclass.
+
+    Remaining keywords are passed into the constructor of the ``dcls``
+
+    Returns
+    -------
+    distr : ``dcls``, usually `Distribution`
+        The sampled Gaussian distribution.
+    """
+    center = np.asanyarray(center)
+    if var is not None:
+        if std is None:
+            std = np.asanyarray(var)**0.5
+        else:
+            raise ValueError('normal cannot take both std and var')
+    if ivar is not None:
+        if std is None:
+            std = np.asanyarray(ivar)**-0.5
+        else:
+            raise ValueError('normal cannot take both ivar and '
+                             'and std or var')
+    if std is None:
+        raise ValueError('normal requires one of std, var, or ivar')
+    else:
+        std = np.asanyarray(std)
+
+    randshape = np.broadcast(std, center).shape + (n_samples,)
+    samples = center[..., np.newaxis] + np.random.randn(*randshape) * std[..., np.newaxis]
+    distr = dcls(samples, **kwargs)
+    distr.distr_std = std
+    return distr
+
+
+def poisson(center, n_samples=1000, dcls=Distribution, **kwargs):
+    """
+    Create a Poisson distribution.
+
+    Parameters
+    ----------
+    center : `~astropy.units.Quantity`
+        The center value of this distribution (i.e., λ).
+    n_samples : int
+        The number of Monte Carlo samples to use with this distribution
+    dcls : class
+        The class to use to create this distribution.  Typically a
+        `Distribution` subclass.
+
+    Remaining keywords are passed into the constructor of the ``dcls``
+
+    Returns
+    -------
+    distr : ``dcls``, usually `Distribution`
+        The sampled poisson distribution.
+    """
+    # we convert to arrays because np.random.poisson has trouble with quantities
+    has_unit = False
+    if hasattr(center, 'unit'):
```

Comment:
I see your point that the example shouldn't use it since kpc is not a counting unit... but I think `electron`/`count`/`adu` are important use cases, particularly given that one of the significant applications here would be for CCDs (via `ccdproc`/`nddata`), which is explicitly made to allow such flexible "counting" units that follow poisson statistics.

So I'd rather be *flexible* here at the outset so that we can experiment.  But perhaps a compromise is to have a warning that is raised for anything other than dimensionless or one of our builtin "counting" units?  I'll give that a try and we'll see how it looks.